cupyx/scipy/signal/_signaltools_core.py

```diff
@@ -92,6 +95,81 @@ def _inputs_swap_needed(mode, shape1, shape2, axes=None):
     not_ok1 = any(shape1[i] < shape2[i] for i in axes)
     not_ok2 = any(shape1[i] > shape2[i] for i in axes)
     if not_ok1 and not_ok2:
-        raise ValueError("For 'valid' mode, one must be at least "
-                         "as large as the other in every dimension")
+        raise ValueError('For "valid" mode, one must be at least '
+                         'as large as the other in every dimension')
     return not_ok1
+
+
+def _init_freq_conv_axes(in1, in2, mode, axes, sorted_axes=False):
+    # See scipy's documentation in scipy.signal.signaltools
+    s1, s2 = in1.shape, in2.shape
+    axes = _init_nd_and_axes(in1, axes)
+    # Length-1 axes can rely on broadcasting rules, no fft needed
+    axes = [a for a in axes if s1[a] != 1 and s2[a] != 1]
+    if sorted_axes:
+        axes.sort()
+
+    # Check that unused axes are either 1 (broadcast) or the same length
+    if not all(s1[a] == s2[a] or s1[a] == 1 or s2[a] == 1
+               for a in range(in1.ndim) if a not in axes):
+        raise ValueError('incompatible shapes for in1 and in2:'
+                         ' {} and {}'.format(s1, s2))
+
+    # Check that input sizes are compatible with 'valid' mode.
+    if _inputs_swap_needed(mode, s1, s2, axes=axes):
+        # Convolution is commutative
+        in1, in2 = in2, in1
+
+    return in1, in2, axes
+
+
+def _init_nd_and_axes(x, axes):
+    # See documentation in scipy.fft._helper._init_nd_shape_and_axes
+    # except shape argument is always None and doesn't return new shape
+    if axes is None:
+        axes = range(x.ndim)
+    else:
+        # Get the list of axes
+        if isinstance(axes, numbers.Number):
```

Comment:
The error message "`axes`  is not iterable" will be shown for a non-number object, but I feel it is confusing.
```suggestion
        if not isinstance(axes, collections.abc.Iterable):
```